src/main/java/duke/parser/Parser.java

```diff
@@ -0,0 +1,166 @@
+package duke.parser;
+
+import java.time.LocalDate;
+import java.time.format.DateTimeParseException;
+import java.util.ArrayList;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import duke.command.Command;
+import duke.command.DeadlineCommand;
+import duke.command.DeleteCommand;
+import duke.command.EventCommand;
+import duke.command.ExitCommand;
+import duke.command.FindCommand;
+import duke.command.HelpCommand;
+import duke.command.ListCommand;
+import duke.command.MarkCommand;
+import duke.command.TodoCommand;
+import duke.command.UnmarkCommand;
+import duke.exception.DukeDatabaseInvalidEntryException;
+import duke.exception.DukeEndDateBeforeStartDateException;
+import duke.exception.DukeInvalidArgumentException;
+import duke.exception.DukeInvalidDateFormatException;
+
+/**
+ * Parses user input and database entries.
+ */
+public class Parser {
+
+    /**
+     * Parses a database entry and extracts relevant information.
+     *
+     * @param entry The database entry to be parsed.
+     * @return An ArrayList containing the elements extracted from the database entry.
+     * @throws DukeDatabaseInvalidEntryException If the database entry format is invalid.
+     */
+    public static ArrayList<String> parseDatabaseEntry(String entry) throws DukeDatabaseInvalidEntryException {
+        ArrayList<String> elements = new ArrayList<>();
+        Pattern pattern = Pattern.compile(
+                "\\[([A-Z])\\]\\[(.)\\] (.+?)(?: \\(by: (.+?)\\)| \\(from: (.+?) to: (.+?)\\))?");
+        Matcher matcher = pattern.matcher(entry);
+
+        if (matcher.matches()) {
+            elements.add(matcher.group(1));
+            elements.add(matcher.group(2));
+            elements.add(matcher.group(3));
+            if (matcher.group(4) != null) {
+                elements.add(matcher.group(4));
+            } else if (matcher.group(5) != null && matcher.group(6) != null) {
+                elements.add(matcher.group(5));
+                elements.add(matcher.group(6));
+            }
+        } else {
+            throw new DukeDatabaseInvalidEntryException();
+        }
+
+        return elements;
+    }
+
+    /**
+     * Parses user input and generates the appropriate command for Duke.
+     *
+     * @param input The user's input command.
+     * @return The corresponding Command object based on the user input.
+     * @throws DukeInvalidArgumentException If the input does not match any valid command format.
+     * @throws DukeInvalidDateFormatException If the input contains an invalid date format.
+     * @throws DukeEndDateBeforeStartDateException If an event's end date is before its start date.
+     */
+    public static Command parseUserInput(String input) throws DukeInvalidArgumentException,
```

Comment:
similarly what is users input are not all lowercase?

Reply:
Yes, you are right, I have added additional format specifiers ie: `String todoPattern = "(?i)todo (.+)";`, where `(?i)` makes the command case insensitive when parsing user input.